### DISCUSSION TITLE
Dynamically laying out Plots in a Grid (or other container?)

### DISCUSSION THREAD

**Discussion by mrpathos** [emilk/egui#4271] [Q&A]
Every time I fight with this I think, "I should post and see if there's a better way to do this" and don't, but not this time!!

I've been using egui for a while now, but struggle with complex layouts. Here's an example I just did:
![image](https://github.com/emilk/egui/assets/66536205/2647ad7d-24e2-4ed7-8f5c-cd1752f8da42)

It took me quite a lot of manual work to get that to all layout that way, and I'm wondering if there's an easier way. In particular, the way I had to do it means if I want to add a few more plots to the grid, I have to go tweak some constants and whatnot, and I'd love a way to just be able to chuck an arbitrary number of these in a grid and have it just divide it all up nicely.

The way I ended up having to do it was:

```
fn render_plot_grid(ui: &mut Ui, sines: &SineHistories) {
    let avail_width = ui.available_width();
    let avail_height = ui.available_height();

    let grid_cell_size = vec2(avail_width / 2.0, avail_height / 4.0);

    ui.spacing_mut().item_spacing = Vec2::ZERO;

    egui::Grid::new("plot grid").show(ui, |ui| {
        draw_named_plot(
            ui,
            grid_cell_size,
            "Sine 1".to_string(),
            sines.sine1.clone(),
        );
        draw_named_plot(
            ui,
            grid_cell_size,
            "Sine 2".to_string(),
            sines.sine2.clone(),
        );
        ui.end_row();
```
and so on, and then `draw_named_plot` does this:
```
fn draw_named_plot(ui: &mut Ui, size: Vec2, bg: Color32, label: String, data: Vec<[f64; 2]>) {
    let inner_margin = vec2(5.0,5.0);
    egui::Frame::none()
        .outer_margin(Margin::ZERO)
        .inner_margin(Margin::symmetric(inner_margin.x, inner_margin.y))
        .fill(bg)
        .show(ui, |ui| {
            ui.set_max_size(size - (inner_margin*2.0));
            ui.with_layout(Layout::top_down(Align::Center), |ui| {
                ui.heading(label.clone());
                Plot::new(label)
                    .show(ui, |plot_ui| plot_ui.line(Line::new(data)));
            });
        });
}
```
What I'd ideally love is to just be able to say something like:

```
Grid::new("my grid").row_ratios([0.5,0.5]).col_ratios([0.25,0.25,0.25,0.25]).show(|ui, index| {
    // this gets called 8 times (or however many rows*cols is), index goes from 0-7, and for each
    // call, ui is a child ui that represents only the space of that grid cell
 }
```
I understand that this can't literally work, because a Grid doesn't automatically consume the entire remaining UI space if the elements you put in don't fill it.

But I am at least wondering if there is a better way than the way I did it, because my approach has a lot of problems:
- I have to manually calculate the size of the grid cells from the ui available size and hardcode the rows and cols there, and it's one more thing to update if I add items
- Plot just attempts to take up all available space, so I don't really know how to put a bunch of them in a grid without explicitly sizing them or the leftmost cells take up all the room and the rightmost ones are just not there.
- egui is full of non-zero default margins, spacings, etc so the explicit size calculation I did then didn't work at all until I spent half an hour going through and doing `outer_margin(Margin::ZERO)`, `inner_margin(Margin::ZERO)`, `ui.spacing_mut().item_spacing = Vec2::ZERO` all over the place
- Even being willing to do that, that process is messy - just keep running and it looks wrong and I have to try to figure out what other automatic spacing I forgot about. I don't, for example, have any idea why there's so much room to the left of each Plot, but I gave up trying to figure that out.

I really love using egui, I should add - and Plot itself is a pretty fantastic tool for diagnostic stuff and visualization.

I'm really just wondering if I'm totally misunderstanding the way I'm supposed to use the layout system. I'd love if I could quickly knock out apps like this and when I want to plot something new, just add another Plot into the Grid's show() having provided some hints like "fill the whole UI and there will be this many rows and cols" to the Grid when I made it, so I don't have to go tweak a bunch of constants and fiddle with it to get it looking right.

Thank you!!

**Comment by YgorSouza** [comment]
I think the [Strip](https://docs.rs/egui_extras/0.27.2/egui_extras/struct.StripBuilder.html) from egui_extras might be a better fit for what you are doing.

<details><summary>Example code</summary>
<p>

```rs
struct MyApp {
    rows: usize,
    columns: usize,
}

impl Default for MyApp {
    fn default() -> Self {
        Self {
            rows: 4,
            columns: 2,
        }
    }
}

impl eframe::App for MyApp {
    fn update(&mut self, ctx: &egui::Context, _frame: &mut eframe::Frame) {
        let Self { rows, columns } = self;
        egui::CentralPanel::default().show(ctx, |ui| {
            ui.horizontal(|ui| {
                ui.label("Rows");
                ui.add(egui::DragValue::new(rows).clamp_range(1..=10));
                ui.label("Columns");
                ui.add(egui::DragValue::new(columns).clamp_range(1..=10));
            });
            ui.spacing_mut().item_spacing = [0.0; 2].into();
            let (rows, columns) = (*rows, *columns);
            egui_extras::StripBuilder::new(ui)
                .sizes(egui_extras::Size::relative((rows as f32).recip()), rows)
                .vertical(|mut strip| {
                    for r in 0..rows {
                        strip.cell(|ui| {
                            egui_extras::StripBuilder::new(ui)
                                .sizes(
                                    egui_extras::Size::relative((columns as f32).recip()),
                                    columns,
                                )
                                .horizontal(|mut strip| {
                                    for c in 0..columns {
                                        strip.cell(|ui| {
                                            let i = r * columns + c;
                                            ui.ctx().debug_painter().debug_rect(
                                                ui.max_rect(),
                                                Color32::GREEN,
                                                format!("{i} ({r}, {c})"),
                                            );
                                        })
                                    }
                                });
                        });
                    }
                });
        });
    }
}
```

</p>
</details> 

[egui_strip_example.webm](https://github.com/emilk/egui/assets/43298013/405013ee-b5d9-4485-8df8-0af983705f81)

Note that I had to set the `item_spacing` to 0 because the strip doesn't seem to handle that very well. You can set the spacing back to the value you want inside each cell's UI.
